core/io/logger.cpp

```diff
@@ -266,3 +267,362 @@ CompositeLogger::~CompositeLogger() {
+		int index_to_send_to = 0;
+
+		while (true) {
+			Callable callable;
+			{
+				MutexLock lock(mutex);
+				if (index_to_send_to >= captures_buffered.size()) {
+					// we done!
+					break;
+				}
+
+				callable = captures_buffered[index_to_send_to++];
+			}
+
+			dispatch_message(log, callable);
+		}
+	}
+
+	if (pre_buffering.is_set()) {
+		// Buffering is done!
+		// Read the comment near the bottom of recalculate_state for a possible race condition and how it's dealt with.
+		// IT IS REALLY IMPORTANT THAT THIS HAPPENS ONLY ON THE MAIN THREAD.
+		MutexLock lock(mutex);
+		pre_buffering.clear();
+		recalculate_state();
+	}
+}
+
+void UserLogManagerLogger::process(const Dictionary &p_message) {
+	Vector<Callable> captures_non_thread_safe_mirror;
+	{
+		// Shove another item into the buffer, if we need to.
+
+		// We also atomically grab `captures_non_thread_safe` to avoid race conditions, in the extraordinarily unlikely case where:
+		// * this function buffers a message, releases the lock, and is immediately preempted
+		// * in another thread, an already-running register_log_capture_non_thread_safe dispatches our new message from the buffer
+		// * that same register_log_capture_non_thread_safe gets to the end of the buffer and adds itself to captures_non_thread_safe
+		// * this function resumes, reaches the captures_non_thread_safe section, and sends the same message again
+
+		// register_log_capture_non_thread_safe intentionally finishes the buffered log message replay atomically with adding a new handler to the vector
+		// so we do the same thing here, adding a message to the log atomically with copying the handlers
+		// Vector COW behavior prevents this from being expensive in the common case
+		MutexLock lock(mutex);
+		if (state.get() == STATE_BUFFERING) {
+			buffered_logs.append(p_message);
+		}
+		captures_non_thread_safe_mirror = captures_non_thread_safe;
+	}
+
+	// Dispatch to all the non_thread_safe callables at the moment we added to the buffer
+	// We actually don't have to care about cutesy thread-safety for once because we're working off a local copy of the captures list
+
+	for (const Callable &callable : captures_non_thread_safe_mirror) {
+		dispatch_message(p_message, callable);
+	}
+}
+
+void UserLogManagerLogger::dispatch_message(const Dictionary &p_message, const Callable &p_callable) {
+	// ideally we should verify that the mutex is *not* held by this thread, but the current API provides no way to do that
+	// (it's possible it's already been scooped up by another thread, that's fine)
+
+	if (!p_callable.is_valid()) {
+		// in most cases, this is going to be a deleted callable, so ignore it;
+		// is_null() would be enough for that, we intentionally invalidate it to null in register_callable
+		//
+		// it's apparently possible for callables to be yanked out from under us, though, and if we call them, we crash
```

Comment:
This comment isn't necessary

Reply:
I do specifically want to call out "it is expected and normal that it's null, because our removal does that"; it's not otherwise clear that this is standard behavior. Do you have another phrasing you'd recommend?

Reply:
No phrasing needed, this is the universal way to check this, it doesn't need explaining here, it's not unusual or unexpected, there's almost no checks against `Callable.is_null` in the codebase, and most remaining ones are invalid and will hopefully be removed soon

This isn't obvious to new users at times, but this isn't the place to clarify this

Reply:
No, again, the thing I want to clarify isn't "we're checking if the callable is valid here", it's "normal functionality of the removal function replaces callables with null callables and this will happen much more often than you would normally expect callable invalidation to happen".

I'll take the entire comment out if you insist but I think that's removing important information.

Reply:
This class is already pretty over annotated, this makes it even more cluttered

But if a comment is necessary just add "Check the callable is valid and hasn't been disconnected", anything more is unnecessary, the code explains itself, it isn't *just* to prevent calling it if it's null, that'll be handled automatically, it's to prevent crashing, but that's something that you don't need to explain here

So anyone reading the code will see that removing the callable sets it to empty, so no confusion is going to happen

Comments should explain things that *aren't* clear from the code, and the clearing is already explained in the other method

Reply:
Generally, once approved, this code needs a comment pass to clean up unnecessary comments and their style as well

Reply:
> This class is already pretty over annotated, this makes it even more cluttered

Intentionally so; it's delicate multithreading work and there's a lot of implicit stuff going on.

> Comments should explain things that aren't clear from the code, and the clearing is already explained in the other method

I think the problem is locality. Sure, the clearing is already explained in the other method, but I can *absolutely* see someone (for example) changing `callp()` to do the `is_valid()` check itself, then looking at every place `callp()` is used and saying "aha, don't need this check anymore!"

If you're looking at A(), and there's an important bit of behavior that you need to know in order to properly modify A(), and that bit of behavior is located in B(), and there's no comment in A() directing you to B(), and there's no other indication that you should go look at B(), then this is - in my opinion - a bug waiting to happen.

> Generally, once approved, this code needs a comment pass to clean up unnecessary comments and their style as well

Is there a way that the reviewer can just submit whatever changes they want? If you want to change the code, you're welcome to just do it.

Again, I think this is a mistake, if you insist I'll change it to whatever you want, but at this point it might just be easier if you made the change instead of me; it'll take less time for you to do it than for you to explain exactly what you want.

Reply:
You don't have to change it, it's just my suggestion, but I'll go over this code once it's approved fully and review the comments, and you can apply those changes then

> but I can absolutely see someone (for example) changing callp() to do the is_valid() check itself

That kind of sweeping change won't happen in such a careless way, and anyone doing so would go over the code carefully, no worries, also this wouldn't make any difference here as the check wouldn't be needed in this case either, so it doesn't really apply here